File: csharp-playwright-framework/PlaywrightFramework/Utilities/TestDataBuilder.cs
using Bogus;

namespace PlaywrightFramework.Utilities
{
    /// <summary>
    /// User data class
    /// </summary>
    public class User
    {
        public string Email { get; set; } = string.Empty;
        public string Password { get; set; } = string.Empty;
        public string FirstName { get; set; } = string.Empty;
        public string LastName { get; set; } = string.Empty;
        public string Phone { get; set; } = string.Empty;
        public string Address { get; set; } = string.Empty;
        public string City { get; set; } = string.Empty;
        public string ZipCode { get; set; } = string.Empty;
        public string Country { get; set; } = string.Empty;
        public string Role { get; set; } = "user";
        public bool IsActive { get; set; } = true;

        public string FullName() => $"{FirstName} {LastName}".Trim();
    }

    /// <summary>
    /// Product data class
    /// </summary>
    public class Product
    {
        public string Name { get; set; } = string.Empty;
        public decimal Price { get; set; }
        public string Description { get; set; } = string.Empty;
        public string Category { get; set; } = string.Empty;
        public string Sku { get; set; } = string.Empty;
        public int Stock { get; set; } = 100;
        public bool IsAvailable { get; set; } = true;
        public string ImageUrl { get; set; } = string.Empty;
    }

    /// <summary>
    /// Order data class
    /// </summary>
    public class Order
    {
        public User User { get; set; } = new User();
        public List<Product> Products { get; set; } = new List<Product>();
        public string OrderId { get; set; } = string.Empty;
        public string Status { get; set; } = "pending";
        public decimal Total { get; set; }
        public string ShippingAddress { get; set; } = string.Empty;
        public string PaymentMethod { get; set; } = "credit_card";

        public decimal CalculateTotal() => Products.Sum(p => p.Price);
    }

    /// <summary>
    /// Builder for creating User test data with fluent interface
    /// </summary>
    /// <example>
    /// var user = new UserBuilder()
    ///     .WithEmail("test@example.com")
    ///     .WithPassword("Test@123")
    ///     .WithName("John", "Doe")
    ///     .WithRole("admin")
    ///     .Build();
    /// </example>
    public class UserBuilder
    {
        private readonly Faker _faker = new Faker();
        private string _email;
        private string _password;
        private string _firstName;
        private string _lastName;
        private string _phone;
        private string _address;
        private string _city;
        private string _zipCode;
        private string _country = "USA";
        private string _role = "user";
        private bool _isActive = true;

        public UserBuilder()
        {
            _email = _faker.Internet.Email();
            _password = DataGenerator.GenerateStrongPassword();
            _firstName = _faker.Name.FirstName();
            _lastName = _faker.Name.LastName();
            _phone = _faker.Phone.PhoneNumber();
            _address = _faker.Address.StreetAddress();
            _city = _faker.Address.City();
            _zipCode = _faker.Address.ZipCode();
        }

        public UserBuilder WithEmail(string email)
        {
            _email = email;
            return this;
        }

        public UserBuilder WithPassword(string password)
        {
            _password = password;
            return this;
        }

        public UserBuilder WithName(string firstName, string lastName)
        {
            _firstName = firstName;
            _lastName = lastName;
            return this;
        }

        public UserBuilder WithPhone(string phone)
        {
            _phone = phone;
            return this;
        }

        public UserBuilder WithAddress(string address, string? city = null, string? zipCode = null, string country = "USA")
        {
            _address = address;
            if (!string.IsNullOrEmpty(city))
                _city = city;
            if (!string.IsNullOrEmpty(zipCode))
                _zipCode = zipCode;
            _country = country;
            return this;
        }

        public UserBuilder WithRole(string role)
        {
            _role = role;
            return this;
        }

        public UserBuilder AsAdmin()
        {
            _role = "admin";
            return this;
        }

        public UserBuilder AsGuest()
        {
            _role = "guest";
            return this;
        }

        public UserBuilder Active(bool isActive = true)
        {
            _isActive = isActive;
            return this;
        }

        public User Build()
        {
            return new User
            {
                Email = _email,
                Password = _password,
                FirstName = _firstName,
                LastName = _lastName,
                Phone = _phone,
                Address = _address,
                City = _city,
                ZipCode = _zipCode,
                Country = _country,
                Role = _role,
                IsActive = _isActive
            };
        }
    }

    /// <summary>
    /// Builder for creating Product test data
    /// </summary>
    /// <example>
    /// var product = new ProductBuilder()
    ///     .WithName("Laptop")
    ///     .WithPrice(999.99m)
    ///     .InCategory("Electronics")
    ///     .WithStock(50)
    ///     .Build();
    /// </example>
    public class ProductBuilder
    {
        private readonly Faker _faker = new Faker();
        private string _name;
        private decimal _price;
        private string _description;
        private string _category = "General";
        private string _sku;
        private int _stock = 100;
        private bool _isAvailable = true;
        private string _imageUrl = "https://via.placeholder.com/300";

        public ProductBuilder()
        {
            _name = $"Product-{_faker.Random.AlphaNumeric(8)}";
            _price = _faker.Random.Decimal(10, 1000);
            _description = _faker.Lorem.Sentence(10);
            _sku = _faker.Random.String2(10, "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789");
        }

        public ProductBuilder WithName(string name)
        {
            _name = name;
            return this;
        }

        public ProductBuilder WithPrice(decimal price)
        {
            _price = price;
            return this;
        }

        public ProductBuilder WithDescription(string description)
        {
            _description = description;
            return this;
        }

        public ProductBuilder InCategory(string category)
        {
            _category = category;
            return this;
        }

        public ProductBuilder WithSku(string sku)
        {
            _sku = sku;
            return this;
        }

        public ProductBuilder WithStock(int stock)
        {
            _stock = stock;
            return this;
        }

        public ProductBuilder OutOfStock()
        {
            _stock = 0;
            _isAvailable = false;
            return this;
        }

        public ProductBuilder WithImage(string imageUrl)
        {
            _imageUrl = imageUrl;
            return this;
        }

        public Product Build()
        {
            return new Product
            {
                Name = _name,
                Price = _price,
                Description = _description,
                Category = _category,
                Sku = _sku,
                Stock = _stock,
                IsAvailable = _isAvailable,
                ImageUrl = _imageUrl
            };
        }
    }

    /// <summary>
    /// Builder for creating Order test data
    /// </summary>
    /// <example>
    /// var order = new OrderBuilder()
    ///     .ForUser(user)
    ///     .WithProducts(new List&lt;Product&gt; { product1, product2 })
    ///     .WithShippingAddress("123 Main St")
    ///     .WithPaymentMethod("paypal")
    ///     .Build();
    /// </example>
    public class OrderBuilder
    {
        private readonly Faker _faker = new Faker();
        private User _user;
        private List<Product> _products = new List<Product>();
        private string _orderId;
        private string _status = "pending";
        private string _shippingAddress = string.Empty;
        private string _paymentMethod = "credit_card";

        public OrderBuilder()
        {
            _user = new UserBuilder().Build();
            _orderId = $"ORD-{_faker.Random.Number(10000000, 99999999)}";
        }

        public OrderBuilder ForUser(User user)
        {
            _user = user;
            return this;
        }

        public OrderBuilder WithProducts(List<Product> products)
        {
            _products = products;
            return this;
        }

        public OrderBuilder AddProduct(Product product)
        {
            _products.Add(product);
            return this;
        }

        public OrderBuilder WithOrderId(string orderId)
        {
            _orderId = orderId;
            return this;
        }

        public OrderBuilder WithStatus(string status)
        {
            _status = status;
            return this;
        }

        public OrderBuilder WithShippingAddress(string address)
        {
            _shippingAddress = address;
            return this;
        }

        public OrderBuilder WithPaymentMethod(string paymentMethod)
        {
            _paymentMethod = paymentMethod;
            return this;
        }

        public Order Build()
        {
            var order = new Order
            {
                User = _user,
                Products = _products,
                OrderId = _orderId,
                Status = _status,
                ShippingAddress = string.IsNullOrEmpty(_shippingAddress) ? _user.Address : _shippingAddress,
                PaymentMethod = _paymentMethod
            };
            order.Total = order.CalculateTotal();
            return order;
        }
    }

    /// <summary>
    /// Convenience methods for common test data patterns
    /// </summary>
    public static class TestDataBuilderHelpers
    {
        public static User CreateStandardUser() => new UserBuilder().Build();

        public static User CreateAdminUser() => new UserBuilder().AsAdmin().Build();

        public static User CreateGuestUser() => new UserBuilder().AsGuest().Build();

        public static Product CreateSampleProduct(string category = "Electronics")
            => new ProductBuilder().InCategory(category).Build();

        public static Order CreateOrderWithProducts(User user, int productCount = 3)
        {
            var products = Enumerable.Range(0, productCount)
                .Select(_ => new ProductBuilder().Build())
                .ToList();
            return new OrderBuilder()
                .ForUser(user)
                .WithProducts(products)
                .Build();
        }
    }
}

File: python-playwright-framework/utils/test_data_builder.py
"""
Test Data Builders using Builder Pattern
Provides fluent interface for creating complex test objects
"""

from typing import Optional
from dataclasses import dataclass, field
from utils.data_generator import DataGenerator


@dataclass
class User:
    """User data class"""
    email: str
    password: str
    first_name: str = ""
    last_name: str = ""
    phone: str = ""
    address: str = ""
    city: str = ""
    zip_code: str = ""
    country: str = ""
    role: str = "user"
    is_active: bool = True
    
    def full_name(self) -> str:
        return f"{self.first_name} {self.last_name}".strip()


@dataclass
class Product:
    """Product data class"""
    name: str
    price: float
    description: str = ""
    category: str = ""
    sku: str = ""
    stock: int = 100
    is_available:bool = True
    image_url: str = ""


@dataclass
class Order:
    """Order data class"""
    user: User
    products: list[Product] = field(default_factory=list)
    order_id: str = ""
    status: str = "pending"
    total: float = 0.0
    shipping_address: str = ""
    payment_method: str = "credit_card"
    
    def calculate_total(self) -> float:
        return sum(p.price for p in self.products)


class UserBuilder:
    """
    Builder for creating User test data with fluent interface
    
    Example:
        user = UserBuilder()\\
            .with_email("test@example.com")\\
            .with_password("Test@123")\\
            .with_name("John", "Doe")\\
            .with_role("admin")\\
            .build()
    """
    
    def __init__(self):
        self._email = DataGenerator.random_email()
        self._password = DataGenerator.random_password()
        self._first_name = DataGenerator.random_first_name()
        self._last_name = DataGenerator.random_last_name()
        self._phone = DataGenerator.random_phone()
        self._address = DataGenerator.random_address()
        self._city = DataGenerator.random_string(10)
        self._zip_code = DataGenerator.random_string(5, "0123456789")
        self._country = "USA"
        self._role = "user"
        self._is_active = True
    
    def with_email(self, email: str):
        """Set email"""
        self._email = email
        return self
    
    def with_password(self, password: str):
        """Set password"""
        self._password = password
        return self
    
    def with_name(self, first_name: str, last_name: str):
        """Set first and last name"""
        self._first_name = first_name
        self._last_name = last_name
        return self
    
    def with_phone(self, phone: str):
        """Set phone number"""
        self._phone = phone
        return self
    
    def with_address(self, address: str, city: str = "", zip_code: str = "", country: str = "USA"):
        """Set address details"""
        self._address = address
        if city:
            self._city = city
        if zip_code:
            self._zip_code = zip_code
        self._country = country
        return self
    
    def with_role(self, role: str):
        """Set user role (user, admin, guest)"""
        self._role = role
        return self
    
    def as_admin(self):
        """Make user an admin"""
        self._role = "admin"
        return self
    
    def as_guest(self):
        """Make user a guest"""
        self._role = "guest"
        return self
    
    def active(self, is_active: bool = True):
        """Set active status"""
        self._is_active = is_active
        return self
    
    def build(self) -> User:
        """Build the User object"""
        return User(
            email=self._email,
            password=self._password,
            first_name=self._first_name,
            last_name=self._last_name,
            phone=self._phone,
            address=self._address,
            city=self._city,
            zip_code=self._zip_code,
            country=self._country,
            role=self._role,
            is_active=self._is_active
        )


class ProductBuilder:
    """
    Builder for creating Product test data
    
    Example:
        product = ProductBuilder()\\
            .with_name("Laptop")\\
            .with_price(999.99)\\
            .in_category("Electronics")\\
            .with_stock(50)\\
            .build()
    """
    
    def __init__(self):
        self._name = f"Product-{DataGenerator.random_string(8)}"
        self._price = DataGenerator.random_number(10, 1000)
        self._description = DataGenerator.random_string(50)
        self._category = "General"
        self._sku = DataGenerator.random_string(10, "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789")
        self._stock = 100
        self._is_available = True
        self._image_url = "https://via.placeholder.com/300"
    
    def with_name(self, name: str):
        """Set product name"""
        self._name = name
        return self
    
    def with_price(self, price: float):
        """Set product price"""
        self._price = price
        return self
    
    def with_description(self, description: str):
        """Set product description"""
        self._description = description
        return self
    
    def in_category(self, category: str):
        """Set product category"""
        self._category = category
        return self
    
    def with_sku(self, sku: str):
        """Set product SKU"""
        self._sku = sku
        return self
    
    def with_stock(self, stock: int):
        """Set stock quantity"""
        self._stock = stock
        return self
    
    def out_of_stock(self):
        """Make product out of stock"""
        self._stock = 0
        self._is_available = False
        return self
    
    def with_image(self, image_url: str):
        """Set product image URL"""
        self._image_url = image_url
        return self
    
    def build(self) -> Product:
        """Build the Product object"""
        return Product(
            name=self._name,
            price=self._price,
            description=self._description,
            category=self._category,
            sku=self._sku,
            stock=self._stock,
            is_available=self._is_available,
            image_url=self._image_url
        )


class OrderBuilder:
    """
    Builder for creating Order test data
    
    Example:
        order = OrderBuilder()\\
            .for_user(user)\\
            .with_products([product1, product2])\\
            .with_shipping_address("123 Main St")\\
            .with_payment_method("paypal")\\
            .build()
    """
    
    def __init__(self):
        self._user = UserBuilder().build()
        self._products = []
        self._order_id = f"ORD-{DataGenerator.random_string(8, '0123456789')}"
        self._status = "pending"
        self._shipping_address = ""
        self._payment_method = "credit_card"
    
    def for_user(self, user: User):
        """Set the user for this order"""
        self._user = user
        return self
    
    def with_products(self, products: list[Product]):
        """Set order products"""
        self._products = products
        return self
    
    def add_product(self, product: Product):
        """Add a single product"""
        self._products.append(product)
        return self
    
    def with_order_id(self, order_id: str):
        """Set order ID"""
        self._order_id = order_id
        return self
    
    def with_status(self, status: str):
        """Set order status (pending, processing, completed, cancelled)"""
        self._status = status
        return self
    
    def with_shipping_address(self, address: str):
        """Set shipping address"""
        self._shipping_address = address
        return self
    
    def with_payment_method(self, payment_method: str):
        """Set payment method"""
        self._payment_method = payment_method
        return self
    
    def build(self) -> Order:
        """Build the Order object"""
        order = Order(
            user=self._user,
            products=self._products,
            order_id=self._order_id,
            status=self._status,
            shipping_address=self._shipping_address or self._user.address,
            payment_method=self._payment_method
        )
        order.total = order.calculate_total()
        return order


# Convenience functions for common patterns
def create_standard_user() -> User:
    """Create a standard user with default values"""
    return UserBuilder().build()


def create_admin_user() -> User:
    """Create an admin user"""
    return UserBuilder().as_admin().build()


def create_guest_user() -> User:
    """Create a guest user"""
    return UserBuilder().as_guest().build()


def create_sample_product(category: str = "Electronics") -> Product:
    """Create a sample product in given category"""
    return ProductBuilder().in_category(category).build()


def create_order_with_products(user: User, product_count: int = 3) -> Order:
    """Create an order with specified number of products"""
    products = [ProductBuilder().build() for _ in range(product_count)]
    return OrderBuilder().for_user(user).with_products(products).build()
